File: go/results.go
// Results: async Result envelopes posted to ResultURL and their parsers.

package mpesa

import (
	"encoding/json"
	"math"
	"regexp"
	"strconv"
	"strings"
)

// AsyncResult is the shared envelope POSTed to ResultURL by async APIs.
type AsyncResult struct {
	Result AsyncResultBody `json:"Result"`
}

// AsyncResultBody carries header fields plus the Key/Value parameter list.
type AsyncResultBody struct {
	ResultType               FlexString        `json:"ResultType"`
	ResultCode               FlexString        `json:"ResultCode"`
	ResultDesc               string            `json:"ResultDesc"`
	OriginatorConversationID string            `json:"OriginatorConversationID"`
	ConversationID           string            `json:"ConversationID"`
	TransactionID            string            `json:"TransactionID"`
	ResultParameters         *ResultParameters `json:"ResultParameters,omitempty"`
	ReferenceData            *ReferenceData    `json:"ReferenceData,omitempty"`
}

// Parameters flattens ResultParameters tolerating absent sections; values are
// rendered leniently as strings since Safaricom mixes types.
//
// First-wins: on duplicate Key entries (gateway retries) the FIRST value
// wins, matching MetadataMap(), Python parameters()/metadata() and
// TypeScript MetadataMap. Earlier last-wins behavior silently diverged
// across SDKs on duplicate keys — fixed to first-wins.
func (r AsyncResultBody) Parameters() map[string]string {
	out := make(map[string]string)
	if r.ResultParameters == nil {
		return out
	}
	for _, p := range r.ResultParameters.ResultParameter {
		if _, exists := out[p.Key]; exists {
			continue
		}
		fs := FlexString("")
		if err := json.Unmarshal(p.Value, &fs); err != nil {
			out[p.Key] = string(p.Value)
			continue
		}
		out[p.Key] = fs.String()
	}
	return out
}

// ResultParameters wraps the repeated {Key,Value} objects.
type ResultParameters struct {
	ResultParameter []ResultParameter `json:"ResultParameter"`
}

// ResultParameter is one async-result key/value pair with a raw value.
type ResultParameter struct {
	Key   string          `json:"Key"`
	Value json.RawMessage `json:"Value,omitempty"`
}

// ReferenceData wraps the optional ReferenceItem echo some async results
// carry (e.g. B2C QueueTimeoutURL).
type ReferenceData struct {
	ReferenceItem []ReferenceItem `json:"ReferenceItem,omitempty"`
}

// UnmarshalJSON tolerates both observed Safaricom shapes: ReferenceItem as a
// single object (b2c.md sample) or as a list.
func (rd *ReferenceData) UnmarshalJSON(b []byte) error {
	var probe struct {
		ReferenceItem json.RawMessage `json:"ReferenceItem"`
	}
	if err := json.Unmarshal(b, &probe); err != nil {
		return err
	}
	if len(probe.ReferenceItem) == 0 || string(probe.ReferenceItem) == "null" {
		return nil
	}
	var single ReferenceItem
	if err := json.Unmarshal(probe.ReferenceItem, &single); err == nil {
		rd.ReferenceItem = []ReferenceItem{single}
		return nil
	}
	var many []ReferenceItem
	if err := json.Unmarshal(probe.ReferenceItem, &many); err != nil {
		return err
	}
	rd.ReferenceItem = many
	return nil
}

// ReferenceItem is one key/value echo entry of async-result ReferenceData.
type ReferenceItem struct {
	Key   string `json:"Key"`
	Value string `json:"Value"`
}

// BalanceSegment is one parsed account row of an Account Balance result.
// The floats are display-only conveniences — Raw preserves the authoritative
// source segment verbatim.
type BalanceSegment struct {
	AccountName string
	Currency    string
	Available   float64
	Uncleared   float64
	Reserved    float64
	Min         float64
	Raw         string
}

// balanceNumRe is the ASCII gate for balance amounts (Python results.py
// _BALANCE_NUM_RE parity): optional sign, 1–18 integer digits, optional
// fraction with 1–6 digits. It rejects what strconv.ParseFloat would
// otherwise accept but must not survive: underscore separators ("1_000"),
// hex/octal ("0x10"), Unicode Nd digits ("١٢٣"), overlong digit runs, and
// empty/alpha strings. The 18-digit cap plus the post-parse 2^53 check keep
// display floats exactly representable.
var balanceNumRe = regexp.MustCompile(`^[+-]?[0-9]{1,18}(\.[0-9]{1,6})?$`)

// maxSafeBalance is 2^53: floats above this lose integer precision, so a
// hostile or corrupt balance blob must be skipped rather than rendered as a
// precision-corrupted display value (Python coerce_amount parity).
const maxSafeBalance = float64(1 << 53)

// maxSafeBalanceStr is the decimal rendering of 2^53 for exact string
// comparison (see exceedsSafeBalance).
const maxSafeBalanceStr = "9007199254740992"

// exceedsSafeBalance reports whether the ASCII-gated decimal text exceeds
// ±2^53 in magnitude WITHOUT rounding through float64 first.
//
// Why string comparison (precision trap): float64 cannot represent
// 9007199254740993 — ParseFloat rounds it to exactly 9007199254740992, so
// a post-parse `math.Abs(v) > 2^53` check passes a hostile value that is
// textually beyond the safe-integer boundary. Comparing the integer digits
// (and any non-zero fraction when equal to max) rejects the hostile text
// exactly. Call only with balanceNumRe-passing input.
func exceedsSafeBalance(text string) bool {
	s := text
	if len(s) > 0 && (s[0] == '+' || s[0] == '-') {
		s = s[1:]
	}
	intPart := s
	fracPart := ""
	if i := strings.IndexByte(s, '.'); i >= 0 {
		intPart = s[:i]
		fracPart = s[i+1:]
	}
	// Strip leading zeros so length comparison is by magnitude.
	intPart = strings.TrimLeft(intPart, "0")
	if intPart == "" {
		intPart = "0"
	}
	if len(intPart) > len(maxSafeBalanceStr) {
		return true
	}
	if len(intPart) < len(maxSafeBalanceStr) {
		return false
	}
	// Same digit count: lexicographic compare equals numeric compare.
	if intPart > maxSafeBalanceStr {
		return true
	}
	if intPart < maxSafeBalanceStr {
		return false
	}
	// Integer part is exactly 2^53: any non-zero fraction exceeds it.
	for i := 0; i < len(fracPart); i++ {
		if fracPart[i] != '0' {
			return true
		}
	}
	return false
}

// ParseBalanceSegments splits the Account Balance result blob — segments
// joined by "&", fields joined by "|" into typed rows. Per account-balance.md
// tolerance requirements, trailing separators, unknown extra fields and
// MALFORMED ROWS ARE SKIPPED AND COUNTED in skipped, never fatal. Parsed
// floats are display-only; BalanceSegment.Raw preserves each source segment.
//
// Amount hardening (N8a, Python results.py parity): each of the four numeric
// fields is first gated by balanceNumRe (ASCII shape), then parsed, then
// rejected if NaN/Inf or beyond ±2^53. This blocks "NaN"/"Inf"/"1e30"
// smuggling through ParseFloat's lenient grammar.
func ParseBalanceSegments(s string) (segments []BalanceSegment, skipped int) {
	for _, seg := range strings.Split(s, "&") {
		fields := strings.Split(seg, "|")
		if len(strings.TrimSpace(strings.Join(fields, ""))) == 0 {
			continue
		}
		row := BalanceSegment{Raw: seg}
		if len(fields) < 6 {
			skipped++
			continue
		}
		row.AccountName = strings.TrimSpace(fields[0])
		row.Currency = strings.TrimSpace(fields[1])
		nums := [4]float64{}
		bad := false
		for i := range nums {
			// ASCII gate BEFORE ParseFloat: float() would otherwise accept
			// "1_000" and Unicode Nd digits that must not survive.
			text := strings.TrimSpace(fields[2+i])
			if !balanceNumRe.MatchString(text) {
				skipped++
				bad = true
				break
			}
			// Precision gate BEFORE float rounding: 9007199254740993
			// parses to exactly 2^53, so the check must be textual.
			if exceedsSafeBalance(text) {
				skipped++
				bad = true
				break
			}
			v, err := strconv.ParseFloat(text, 64)
			if err != nil {
				skipped++
				bad = true
				break
			}
			// Non-finite (NaN/±Inf) defense-in-depth: the ASCII gate
			// already blocks these spellings, but a future regex change
			// must never surface them as display values.
			if math.IsNaN(v) || math.IsInf(v, 0) || math.Abs(v) > maxSafeBalance {
				skipped++
				bad = true
				break
			}
			nums[i] = v
		}
		if bad {
			continue
		}
		row.Available, row.Uncleared, row.Reserved, row.Min = nums[0], nums[1], nums[2], nums[3]
		segments = append(segments, row)
	}
	return segments, skipped
}

File: python/mpesa/results.py
"""Async ``Result`` envelopes POSTed to ``ResultURL`` -- the FINAL verdicts
of B2C, Transaction Status, Reversal and Account Balance
(mirrors go/results.go).

Reminder: the earlier ``ConversationResponse`` was merely the queue ACK;
this envelope carries the actual outcome. Like every Safaricom push it
is UNSIGNED -- bind on ``originator_conversation_id`` against your own
records before acting, cap ingestion body size at the framework level,
and classify per ADR-010 (INDETERMINATE may still be retried via query).

Usage::

    result = AsyncResult.from_json(raw_body)
    if result.classify() is ResultClass.SUCCESS:
        receipt = result.transaction_receipt()
        funds = result.amount()
"""

from __future__ import annotations

import json
import math
import re
from dataclasses import dataclass, field
from typing import Any

from .classification import ResultClass, classify_result_code
from .coercion import coerce_amount, coerce_int, coerce_str, first_wins, safe_json_int

__all__ = ["AsyncResult", "Parameter", "ReferenceItem", "BalanceSegment",
           "parse_balance_segments"]

# Ingestion cap in BYTES (Go maxResponseLen / TS MAX_RESPONSE_LEN parity).
# str bodies are measured as UTF-8 bytes: 1M CJK chars ~= 3 MiB and must
# not bypass the bound. See _check_body_size().
_MAX_BODY_BYTES = 1_048_576
_BALANCE_NUM_RE = re.compile(r"[+-]?[0-9]{1,18}(\.[0-9]{1,6})?", re.ASCII)


def _check_body_size(data: "bytes | bytearray | str") -> None:
    """Reject bodies over the ingestion cap, measured in UTF-8 bytes.

    bytes input is measured directly; str input is measured as
    ``len(data.encode("utf-8"))`` so multi-byte payloads cannot smuggle
    up to 4x the intended bound past a char-count check.
    """
    size = len(data) if isinstance(data, (bytes, bytearray)) else \
        len(data.encode("utf-8"))
    if size > _MAX_BODY_BYTES:
        raise ValueError(
            f"mpesa: result body exceeds {_MAX_BODY_BYTES} bytes")


@dataclass(frozen=True)
class Parameter:
    """One async-result Key/Value pair; Value kept as decoded JSON."""

    key: str
    value_raw: Any = None


@dataclass(frozen=True)
class ReferenceItem:
    """One echoed key/value entry of optional ``ReferenceData``."""

    key: str
    value: str = ""


@dataclass(frozen=True)
class BalanceSegment:
    """One parsed Account Balance account row. Floats are DISPLAY-ONLY
    conveniences -- ``raw`` preserves the authoritative source verbatim."""

    account_name: str = ""
    currency: str = ""
    available: float = 0.0
    uncleared: float = 0.0
    reserved: float = 0.0
    min_amount: float = 0.0
    raw: str = ""


def parse_balance_segments(
        text: str) -> "tuple[tuple[BalanceSegment, ...], int]":
    """Split the balance blob (segments joined by '&', fields by '|').

    Tolerates trailing separators and unknown extra fields; MALFORMED
    ROWS ARE SKIPPED AND COUNTED, never fatal (docs/apis/account-balance.md).

    Example::

        segments, skipped = parse_balance_segments(result_text)
    """
    segments: list[BalanceSegment] = []
    skipped = 0
    for seg in text.split("&"):
        fields = seg.split("|")
        if not "".join(fields).strip():
            continue  # blank gap between separators: ignored, not counted
        if len(fields) < 6:
            skipped += 1
            continue
        # ASCII gate: float() would otherwise accept "1_000" and Unicode
        # Nd digits ("١٢٣") that Go's ParseFloat rejects.
        if not all(_BALANCE_NUM_RE.fullmatch(f.strip()) for f in fields[2:6]):
            skipped += 1
            continue
        try:
            nums = [float(f.strip()) for f in fields[2:6]]
        except ValueError:
            skipped += 1
            continue
        if not all(math.isfinite(n) for n in nums):
            skipped += 1
            continue
        segments.append(BalanceSegment(
            account_name=fields[0].strip(), currency=fields[1].strip(),
            available=nums[0], uncleared=nums[1], reserved=nums[2],
            min_amount=nums[3], raw=seg))
    return tuple(segments), skipped


@dataclass(frozen=True)
class AsyncResult:
    """The shared async-result envelope. ``result_code`` normalized to
    str; parameters/reference sections tolerate absence."""

    result_type: int | None = None
    result_code: str = ""
    result_desc: str = ""
    originator_conversation_id: str = ""
    conversation_id: str = ""
    transaction_id: str = ""
    _parameters: tuple[Parameter, ...] = field(default=(), repr=False)
    _reference_items: tuple[ReferenceItem, ...] = field(default=(), repr=False)

    @classmethod
    def from_json(cls, data: "dict | bytes | str") -> "AsyncResult":
        """Parse ``{"Result": {...}}`` loudly about SHAPE (dotted-path
        errors naming every missing scalar), tolerant about TYPES.
        Bodies over 1 MiB (UTF-8 bytes) are rejected before parsing."""
        if isinstance(data, (bytes, bytearray, str)):
            _check_body_size(data)
        if isinstance(data, (bytes, bytearray)):
            data = data.decode("utf-8", errors="replace")
        if isinstance(data, str):
            try:
                data = json.loads(data, parse_int=safe_json_int)
            except (ValueError, RecursionError) as exc:
                raise ValueError(
                    f"mpesa: unparseable result body "
                    f"({type(exc).__name__})") from None
        if not isinstance(data, dict):
            raise ValueError("mpesa: unexpected AsyncResult shape: "
                             "missing or malformed Result")
        inner = data.get("Result")
        if not isinstance(inner, dict):
            raise ValueError("mpesa: unexpected AsyncResult shape: "
                             "missing or malformed Result")
        wire = (("result_type", "ResultType"), ("result_code", "ResultCode"),
                ("result_desc", "ResultDesc"),
                ("originator_conversation_id", "OriginatorConversationID"),
                ("conversation_id", "ConversationID"),
                ("transaction_id", "TransactionID"))
        missing = [key for _, key in wire if key not in inner]
        if missing:
            raise ValueError(f"mpesa: unexpected AsyncResult shape: "
                             f"missing {', '.join(missing)}")
        kwargs = {attr: (coerce_int(inner[key]) if attr == "result_type"
                         else coerce_str(inner[key]) or "")
                  for attr, key in wire}
        params = inner.get("ResultParameters") or {}
        plist = params.get("ResultParameter") if isinstance(params, dict) else None
        items = tuple(Parameter(key=str(p.get("Key", "")),
                                value_raw=p.get("Value"))
                      for p in (plist or []) if isinstance(p, dict))
        ref = inner.get("ReferenceData") or {}
        rlist = ref.get("ReferenceItem") if isinstance(ref, dict) else None
        if isinstance(rlist, dict):  # single-object shape (b2c.md sample)
            rlist = [rlist]
        refs = tuple(ReferenceItem(key=str(r.get("Key", "")),
                                   value=coerce_str(r.get("Value")) or "")
                     for r in (rlist or []) if isinstance(r, dict))
        return cls(_parameters=items, _reference_items=refs, **kwargs)

    def parameters(self) -> dict[str, Any]:
        """Flatten Key/Value pairs FIRST-WINS; absent section yields {}."""
        return first_wins([(p.key, p.value_raw) for p in self._parameters])

    def reference_items(self) -> tuple[ReferenceItem, ...]:
        """Echoed ReferenceItem entries (single-object shapes merged)."""
        return self._reference_items

    def classify(self) -> ResultClass:
        """ADR-010 bucket for ``result_code`` (never auto-fail unknowns)."""
        return classify_result_code(self.result_code)

    def _param(self, name: str) -> Any:
        for parameter in self._parameters:
            if parameter.key == name:
                return parameter.value_raw
        return None

    def transaction_receipt(self) -> str | None:
        """TransactionReceipt string, or None when absent."""
        return coerce_str(self._param("TransactionReceipt"))

    def transaction_status(self) -> str | None:
        """TransactionStatus string (status APIs), or None when absent."""
        return coerce_str(self._param("TransactionStatus"))

    def amount(self) -> float | None:
        """TransactionAmount as float via shared :func:`coerce_amount`;
        bools, >2**53 ints, non-finite floats and non-decimal strings
        yield None (hostile-int guards)."""
        return coerce_amount(self._param("TransactionAmount"))
